### PR TITLE
Associated detectors flyout through UI actions

### PR DESCRIPTION
**Pull request by amitgalitz**
### Description

This change adds the associated detectors flyout as part of the feature-anywhere project. This flyout allows you to view all the detectors associated with a particular visualization. From the flyout a user can unlink a detector from a visualization or search up their associated detectors and click on it to view the detector details on the AD plugin.

In this PR the "Associate Detector" button doesn't do anything, the reason for this is that it relies on some other changes in the CreateAnomalyDetector flyout that aren't merged yet. I will add this functionality on another PR.

This code also might have some build failures because it relies on files such as [public/action/ad_dashboard_action.tsx](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/442/files#diff-973cb51dfb9b0c8247635668d26f0988132cb593eba59e2aa7983c47736bd7c4R2) to run correctly. The reason this file as well as some of the CreateAnomalyDetector files aren't included is because it will be merged first by Jackie's PR. 

This PR also requires additional testing that I will be adding. 

Screenshot of flyout:
Screenshot 2023-03-27 at 10 00 21 AM

### Check List
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:43]: Whats the reason for the extra space (`&nbsp;`) here?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:48]: Maybe this is from UX, but this whole sentence seems grammatically weird. I would prefer something like
`Removing the association unlinks detector '<detector-name>' from the visualization, but does not delete it`

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:62]: nit - could we change the loading state to just set `isLoading={true}` for the button, instead of having no button at all? In other words, remove the conditional and just populate the button's `isLoading` prop with the value of the existing `isLoading` prop that you are passing in.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:38]: nit - you can add `index.ts` files for each directory and sub-directory to handle the exports. This way it simplifies imports since you just have to reference directories instead of the implemented file. e.g., changing the above to 
```
import { ConfirmUnlinkDetectorModal } from '../components';
```

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:125]: You should be able to pass some arguments to `findAll()` to handle filtering for you ([ref](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/feature/feature-anywhere/src/plugins/saved_objects/public/saved_object/saved_object_loader.ts#L131)). For example, passing in search fields on `originPlugin` and/or `visId` with the values you want. This way you wouldn't need to have all of the helper fns.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:198]: 2 things:
1. Should the first sentence be something like "The detector's anomalies may still appear on the visualization"? Because from the second sentence, it seems that's the case
2. Was it determined from UX that we don't want to auto-refresh for the user?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:254]: When would the stored detector list item be empty?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:320]: nit - would it be cleaner to just instantiate the confirm modal jsx here instead of embedding within a function to fetch it?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:336]: Should we add an `isLoading` prop to the button here so that users cant click on it if there is still data being loaded?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/styles.scss
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/styles.scss:6]: Can this import be removed?

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/Container/Container.tsx:5]: nit - can we have this named something else to be more specific here? `Container` is a bit vague.

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/Container/Container.tsx:7]: we should handle `index` better such that if it doesn't exist, it can be passed down to `Flyout` component with that information. At least specifying the type as `string | undefined` can be helpful here

File: public/components/FeatureAnywhereContextMenu/Container/styles.scss
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/Container/styles.scss:4]: I believe this is copied over from here: https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/450#discussion_r1154971204

But pointing out my concerns with that here as well

File: public/utils/contextMenu/styles.scss
Comment (ohltyler) [inline public/utils/contextMenu/styles.scss:6]: what is the purpose of this import?

File: public/components/FeatureAnywhereContextMenu/Container/styles.scss
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/Container/styles.scss:4]: Yes I actually agree, I think we should use small or medium. I missed this while trying to align with Jackie's code. I have the associated detectors style set to medium for the flyout, will check with UX if its okay to go with that for all ad flyouts

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:336]: Hmm I don't think in this case we need an isLoading since this is just to send them to the associate with existing detector flyout. It wont associate a detector so it has no direct dependency to the data being loaded. The only use case I can see is that if user clicks on this cause they see they don't have any detectors currently associated with the viz but that would have a loading component there. Users might want to still go to associate an existing detector flyout while all the current ones are loading. I can bring this up with UX team though.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:48]: yes the wording here was given by UX. I see your point, I will bring it up with UX/techinical writers and fix up the wording here.

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/Container/Container.tsx:7]: Looking into this, not as easy as `string | undefined` as right now it returns '{ label: any; }[]'

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/Container/Container.tsx:5]: Does `AnywhereParentFlyout` sounds good?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:125]: Would you prefer that this is used then what I already have coded, just in terms I know this already works and has been tested. We can also have any future optimization or changes in our control here.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:198]: 1. This makes sense to me, I'll confirm with UX as they gave wording but it does seem like a logic error.
2. Yes currently UX is okay with not auto-refreshing, also both alerting folks and I haven't been able to get the auto refresh of the dashboard to work even if we wanted it.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:254]: I actually don't think it would be possible on a second thought here since its coming directly from the list of detectors. I think originally I was thinking that there would be some lag if detector was already unlinked but I guess it wouldn't be in this operation cause its coming from the stored list and could maybe happen on the crud operations where i do have error checking

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:320]: I can do this, just wanted to a shorted component here but no strong preference

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:48]: Sounds good. Not a blocker to me - would be great to follow up.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/components/ConfirmUnlinkDetectorModal/ConfirmUnlinkDetectorModal.tsx:62]: Got it - my initial thought was this would be in loading state when opening it, instead of closing/finishing the action. Makes sense to me to leave as-is.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:125]: I would prefer the way I recommended since it would allow us to leverage OpenSearch and be more optimized and simpler. Also, it would be less code we have to maintain or test.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:198]: Got it.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:254]: Cool - let's go ahead and remove it

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/Container/Container.tsx:5]: fine with me :)

File: public/components/FeatureAnywhereContextMenu/Container/Container.tsx
Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/Container/Container.tsx:7]: Right - or perhaps `{ label: string | undefined }`, or set default as empty string:
```
const index = [{ label: get(embeddable, 'vis.data.indexPattern.title', '') }];
```

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:125]: Discussed offline, filtering through FindAll is limited to title and description, issue is here: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3566

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:38]: I added a few index.ts to handle exports, should we do this even for every components folder that just has one file? didn't want to cause way too many files.

File: public/utils/contextMenu/getActions.tsx
Comment (ohltyler) [inline public/utils/contextMenu/getActions.tsx:85]: Let's leverage BASE_DOCS_LINK here: https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/utils/constants.ts#L54

Comment (ohltyler) [reply public/utils/contextMenu/getActions.tsx:85]: And if the docs link needs to be updated, can you help update?

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx]: Added this

File: public/utils/contextMenu/getActions.tsx
Comment (amitgalitz) [reply public/utils/contextMenu/getActions.tsx]: Changed to use the doc link constant. No need to update here actually.

**Comment by ohltyler** [comment]
LGTM! Will wait to fully approve when rebased after #482 is merged first.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:193]: nit: you could combine all of these into a chained set of cmds to simplify it and to remove having to keep intermediate consts.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:165]: `ISavedAugmentVis` interface has changed - see [here](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4006/files#diff-f3a7084285f53c21fad7d46ef7e11a91926e5568f3ac0c9ef9df96f015f99f2a).

Basically this will mean to change above to:
```
get(savedObject, 'pluginResource.id', '')
```

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (ohltyler) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:238]: Is this pending once create detector flyout is merged first?

Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:238]: and this: https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/484

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:235]: No this will be finished with this "associate-existing" PR I commented below

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:273]: This is to catch if there is anything in the search box, if we are rendering the empty message we need to know if its because there is no selected detectors in the state at all or if its because the user's search result is returning nothing. The message would then differ with something like "There are no detectors matching your search" vs "There are no anomaly detectors associated with ${props.embeddableTitle} visualization."

Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:273]: The renderEmptyMessage() is automatically called by the EuiInMemoryTable when there is no data to display in table

Comment (ohltyler) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:273]: Got it, that makes sense

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (amitgalitz) [reply public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx]: I can delete savedAugmentObjects and savedObjectToUnlinkId. Wanted to keep 2 constants for the filter operations to keep them more readable rather then nesting the gets and filters too much.

**Comment by ohltyler** [comment]
Cool, this lgtm. Won't approve until dependent PRs merged

**Review by ohltyler — DISMISSED** [review]

File: public/components/FeatureAnywhereContextMenu/AnywhereParentFlyout/AnywhereParentFlyout.tsx
Comment (jackiehanyang) [inline public/components/FeatureAnywhereContextMenu/AnywhereParentFlyout/AnywhereParentFlyout.tsx:20]: we shouldn't remove this line, instead, please append the `associated: AssociatedDetectors,` as the second line here

File: public/components/FeatureAnywhereContextMenu/AnywhereParentFlyout/index.ts
Comment (jackiehanyang) [inline public/components/FeatureAnywhereContextMenu/AnywhereParentFlyout/index.ts:1]: this file is not needed, we already have `index.tsx` file.

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (jackiehanyang) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:50]: Please avoid passing core as a prop around as tyler suggested in my previous pr - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/482#discussion_r1198345354

File: public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx
Comment (jackiehanyang) [inline public/components/FeatureAnywhereContextMenu/AssociatedDetectors/containers/AssociatedDetectors.tsx:90]: same as above, please remove `core.notifications` usage, instead, just get the notification service to use here

File: public/components/FeatureAnywhereContextMenu/DocumentationTitle/index.ts
Comment (jackiehanyang) [inline public/components/FeatureAnywhereContextMenu/DocumentationTitle/index.ts:1]: this file can be removed

File: public/utils/constants.ts
Comment (jackiehanyang) [inline public/utils/constants.ts:56]: you might want to rebase the branch, this constant has already been added - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/featureAnywhere/public/utils/constants.ts#L56

Comment (amitgalitz) [reply public/utils/constants.ts:56]: I think this is just running prettier on the line

**Review by jackiehanyang — APPROVED** [review]

**Review by ohltyler — APPROVED** [review]

**Pull request merged**
